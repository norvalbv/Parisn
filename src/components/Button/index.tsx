'use client';

import { type VariantProps, cva } from 'class-variance-authority';
import type { LucideIcon } from 'lucide-react';
import React, { useState, type ReactElement } from 'react';
import Loading from '../Loading';
import { cn } from '@/lib/utils/cn';

const buttonVariants = cva(
  'inline-flex items-center justify-center transition-all duration-200 focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-text-primary disabled:pointer-events-none disabled:opacity-50',
  {
    variants: {
      variant: {
        dark: 'text-text-primary bg-background hover:bg-zinc-800/80',
        ghost: 'text-text-primary bg-transparent hover:bg-zinc-800/20',
      },
      size: {
        xs: 'px-2 py-1.5 text-xs sm:px-3 sm:py-2',
        sm: 'px-3 py-2 text-xs sm:px-4 sm:py-2.5',
        md: 'px-4 py-2.5 text-sm sm:px-6 sm:py-3',
        lg: 'px-6 py-3 text-sm sm:px-8 sm:py-3.5 sm:text-base',
        icon: 'h-9 w-9',
      },
      fullWidth: {
        true: 'w-full',
        false: 'w-fit',
      },
      hoveredAnimation: {
        true: 'hover:scale-[1.02]',
        false: '',
      },
    },
    defaultVariants: {
      variant: 'dark',
      size: 'md',
      fullWidth: false,
      hoveredAnimation: false,
    },
  }
);

type IconType = LucideIcon | React.ReactElement;

type Props = React.ButtonHTMLAttributes<HTMLButtonElement> &
  VariantProps<typeof buttonVariants> & {
    icon?: IconType;
    iconPosition?: 'left' | 'right';
    loading?: boolean;
    hoveredText?: string;
  };

const Button = ({
  className,
  variant,
  size,
  fullWidth,
  hoveredAnimation,
  icon: Icon,
  iconPosition = 'left',
  loading,
  hoveredText,
  children,
  ...props
}: Props): ReactElement => {
  const [hovered, setHovered] = useState(false);

  const renderIcon = (position: 'left' | 'right'): ReactElement | null => {
    if (!Icon) return null;

    const iconClass = `${position === 'left' ? 'mr-2' : 'ml-2'} size-6`;

    if (React.isValidElement(Icon)) {
      return React.cloneElement(Icon as ReactElement, { className: iconClass });
    }

    const LucideIcon = Icon as LucideIcon;
    return <LucideIcon strokeWidth={1} className={iconClass} />;
  };

  return (
    <button
      className={cn(
        buttonVariants({
          variant,
          size,
          fullWidth,
          hoveredAnimation,
          className,
        })
      )}
      onMouseEnter={(): void => setHovered(true)}
      onMouseLeave={(): void => setHovered(false)}
      disabled={loading}
      {...props}
    >
      {iconPosition === 'left' && renderIcon('left')}
      <span className="inline-block whitespace-nowrap tracking-wider uppercase">
        {hovered && hoveredText ? hoveredText : children}
      </span>
      {loading && <Loading className="ml-2" />}
      {iconPosition === 'right' && renderIcon('right')}
    </button>
  );
};

export default Button;
