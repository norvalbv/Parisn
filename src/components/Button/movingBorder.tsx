'use client';

import {
  motion,
  useAnimationFrame,
  useMotionTemplate,
  useMotionValue,
  useTransform,
} from 'framer-motion';
import React, { useRef } from 'react';

export const MovingBorder = ({
  children,
  duration = 2000,
  rx,
  ry,
  ...otherProps
}: {
  children: React.ReactNode;
  duration?: number;
  rx?: string;
  ry?: string;
  [key: string]: any;
}) => {
  const pathRef = useRef<any>();
  const progress = useMotionValue<number>(0);

  useAnimationFrame((time) => {
    const length = pathRef.current?.getTotalLength();
    if (length) {
      const pxPerMillisecond = length / duration;
      progress.set((time * pxPerMillisecond) % length);
    }
  });

  const x = useTransform(progress, (val) => pathRef.current?.getPointAtLength(val).x);
  const y = useTransform(progress, (val) => pathRef.current?.getPointAtLength(val).y);

  const transform = useMotionTemplate`translateX(${x}px) translateY(${y}px) translateX(-50%) translateY(-50%)`;

  return (
    <>
      <svg
        xmlns="http://www.w3.org/2000/svg"
        preserveAspectRatio="none"
        className="absolute h-full w-full"
        width="100%"
        height="100%"
        {...otherProps}
      >
        <rect fill="none" width="100%" height="100%" rx={rx} ry={ry} ref={pathRef} />
      </svg>
      <motion.div
        style={{
          position: 'absolute',
          top: 0,
          left: 0,
          display: 'inline-block',
          transform,
        }}
      >
        {children}
      </motion.div>
    </>
  );
};
