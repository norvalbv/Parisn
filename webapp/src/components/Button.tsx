import React, { ReactElement, forwardRef } from "react";
import Spinner from "./Spinner";

interface ButtonProps {
  /**
   * Text or label of button
   */
  text: string;
  /**
   * How big the button
   */
  size?: keyof typeof buttonSizeMap;
  /**
   * How big the border radius
   */
  rounded?: keyof typeof buttonRadiusMap;
  /**
   * Set button width manually
   */
  width?: number;
  /**
   * Overwrite text color
   */
  color?: string;
  /**
   * Overwrite background color
   */
  backgroundColor?: string;
  /**
   * Is button disabled
   */
  disabled?: boolean;
  /**
   * Loading state
   */
  loading?: boolean;
  /**
   * Icon for button
   */
  icon?: JSX.Element;
  /**
   * Icon position
   */
  iconPosition?: "left" | "right";
  /**
   * Optional click event
   */
  onClick?: () => void;
  /**
   * id to identify button on dom
   */
  id?: string;
  /*
   * Data attribute for testing with cypress
   */
  dataAtt?: string;
  /**
   * Text case
   */
  upperCase?: boolean;
  /**
   * Additional classes
   */
  classes?: string;
  /**
   * Font weight for text
   */
  fontWeight?: keyof typeof fontWeightMap;
}

const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  (
    {
      text,
      size = "base",
      rounded = "none",
      width = "25rem",
      color,
      backgroundColor,
      disabled,
      loading = false,
      icon,
      iconPosition = "left",
      onClick,
      id,
      dataAtt,
      upperCase = true,
      classes = "",
      fontWeight = "semibold",
    },
    ref
  ): ReactElement => {
    const clickHandle = (): void => {
      if (loading) {
        return;
      }

      if (onClick) onClick();
    };

    return (
      <button
        type="submit"
        className={`relative inline-flex items-center justify-center text-center hover:bg-buttons-hover ${
          upperCase ? "uppercase" : ""
        } ${buttonRadiusMap[rounded]} ${
          buttonSizeMap[size]
        } ${classes} border py-4`}
        style={{ width, color, backgroundColor }}
        onClick={(): void => clickHandle()}
        role="button"
        id={id}
        data-testid={dataAtt}
        disabled={disabled}
        ref={ref}
      >
        {disabled && (
          <div className="absolute top-0 left-0 h-full w-full bg-white opacity-50" />
        )}
        {loading && (
          <div className="absolute top-0 left-0 h-full w-full bg-white opacity-50">
            {text}
            <span className="float-right mt-3">
              <Spinner />
            </span>
          </div>
        )}
        {icon && iconPosition === "left" && (
          <span className="mr-2.5">{icon}</span>
        )}
        <span className={`inline-block ${fontWeightMap[fontWeight]}`}>
          {text}
        </span>
        {icon && iconPosition === "right" && <span>{icon}</span>}
      </button>
    );
  }
);

Button.displayName = "Button";

const buttonSizeMap = {
  xs: "px-1 py-2 text-sm",
  sm: "px-5 py-2.5 text-sm",
  base: "px-6 py-2.5 text-sm w-full",
  lg: "px-8 py-2.5 text-base w-full",
};

const buttonRadiusMap = {
  xs: "rounded-xs",
  small: "rounded-small",
  base: "rounded",
  md: "rounded-md",
  lg: "rounded-lg",
  xl: "rounded-xl",
  none: "",
};

const fontWeightMap = {
  thin: "font-thin",
  extralight: "font-extralight",
  light: "font-light",
  normal: "font-normal",
  medium: "font-medium",
  semibold: "font-semibold",
  bold: "font-bold",
  extrabold: "font-extrabold",
  black: "font-black",
};

export default React.memo(Button);
