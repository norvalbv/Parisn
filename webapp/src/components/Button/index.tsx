import React, { ReactElement, forwardRef, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import {
  BorderRequired,
  borderRequiredMap,
  FontWeight,
  fontWeightMap,
  Rounded,
  roundedMap,
} from '../../types/tailwind';
import { Spinner } from '../CustomSVG';

export interface ButtonProps {
  /**
   * Text or label of button
   */
  text: string | JSX.Element;
  /**
   * Text or label for when hovered
   */
  hoveredText?: string;
  /**
   * How big the button is
   */
  size?: keyof typeof buttonSizeMap;
  /**
   * How big the border radius
   */
  rounded?: Rounded;
  /**
   * Set button width manually
   */
  width?: string;
  /**
   * Overwrite text color
   */
  color?: string;
  /**
   * Overwrite background color
   */
  backgroundColor?: string;
  /**
   * Require a hover colour?
   */
  hoverColorRequired?: boolean;
  /**
   * Is button disabled
   */
  disabled?: boolean;
  /**
   * Loading state
   */
  loading?: boolean;
  /**
   * Icon for button
   */
  icon?: JSX.Element;
  /**
   * Icon position
   */
  iconPosition?: 'left' | 'right';
  /**
   * Optional click event
   */
  onClick?: (arg1?: unknown) => void;
  /**
   * Optional mouse event
   */
  onMouseLeave?: () => void;
  /**
   * id to identify button on dom
   */
  id?: string;
  /*
   * Data attribute for testing with cypress
   */
  dataAtt?: string;
  /**
   * Text case
   */
  upperCase?: boolean;
  /**
   * Additional classes
   */
  classes?: string;
  /**
   * Positioning of button
   */
  positioning?: 'relative' | 'absolute';
  /**
   * Font weight for text
   */
  fontWeight?: FontWeight;
  /**
   * Type of button
   */
  type?: 'button' | 'submit';
  /**
   * Navigate to another page within the app
   */
  navigateTo?: string;
  /**
   * Require boarders?
   */
  borderRequired?: BorderRequired;
  /**
   * Positioning of text
   */
  textOrientation?: string;
}

const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  (
    {
      backgroundColor,
      borderRequired = 'all',
      classes = '',
      color,
      dataAtt,
      disabled,
      fontWeight = 'semibold',
      hoverColorRequired = true,
      hoveredText,
      icon,
      iconPosition = 'left',
      id,
      loading = false,
      navigateTo,
      onClick,
      onMouseLeave,
      positioning = 'relative',
      rounded = 'none',
      size = 'base',
      text,
      type = 'button',
      upperCase = true,
      width = '25rem',
      textOrientation = 'justify-center',
    },
    ref
  ): ReactElement => {
    const [hovered, setHovered] = useState(false);

    const navigate = useNavigate();
    const clickHandle = (): void => {
      if (loading) {
        return;
      }

      if (onClick) onClick();

      if (navigateTo) navigate(navigateTo);
    };

    return (
      <button
        type={type}
        className={`${positioning} inline-flex items-center ${textOrientation} ${
          hoverColorRequired && 'hover:bg-buttons-hover transition-colors duration-200'
        } ${upperCase ? 'uppercase' : ''} ${roundedMap[rounded]} ${
          buttonSizeMap[size]
        } ${classes} ${borderRequiredMap[borderRequired]} py-4`}
        style={{ width, color, backgroundColor }}
        onClick={(): void => clickHandle()}
        role="button"
        id={id}
        data-testid={dataAtt}
        disabled={disabled}
        ref={ref}
        onMouseEnter={() => setHovered(true)}
        onMouseLeave={() => {
          setHovered(false);
          if (onMouseLeave) onMouseLeave();
        }}
      >
        {disabled && (
          <svg width="100%" height="100%" className="absolute">
            <line x2="100%" y2="100%" stroke="darkRed" strokeWidth={2.25} />
          </svg>
        )}
        {loading && (
          <div className="absolute top-0 left-0 h-full w-full bg-white opacity-50">
            {text}
            <span className="float-right mt-3">
              <Spinner />
            </span>
          </div>
        )}
        {icon && iconPosition === 'left' && <span className="mr-2.5">{icon}</span>}
        <span className={`inline-block ${fontWeightMap[fontWeight]}`}>
          {hovered && hoveredText ? hoveredText : text}
        </span>
        {icon && iconPosition === 'right' && <span>{icon}</span>}
      </button>
    );
  }
);

Button.displayName = 'Button';

const buttonSizeMap = {
  xs: 'px-1 py-2 text-sm',
  sm: 'px-5 py-2.5 text-sm',
  base: 'px-6 py-2.5 text-sm w-full',
  lg: 'px-8 py-2.5 text-base w-full',
};

export default React.memo(Button);
