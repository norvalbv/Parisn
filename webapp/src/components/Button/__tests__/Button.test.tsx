import { FC } from 'react';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';

import Button, { ButtonProps } from '../';

const ButtonWithRequiredProps: FC<Required<ButtonProps>> = Button;

describe('<Button />', () => {
  test('Button props with mandatory props', () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <Button text="Test Text" />
        </BrowserRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Button props with all props as required', () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <ButtonWithRequiredProps
            text="Test Text"
            hoveredText="Hovered Text"
            size="base"
            rounded="none"
            width="25rem"
            color="text-red-500"
            backgroundColor="bg-red-200"
            hoverColorRequired
            disabled
            loading={false}
            icon={<div>Icon</div>}
            iconPosition="left"
            onClick={jest.fn()}
            onMouseLeave={jest.fn()}
            id="Test Id"
            dataAtt="Data Attribute"
            upperCase
            classes="underline"
            fontWeight="semibold"
            type="button"
            navigateTo="/"
            borderRequired="all"
            positioning="relative"
            textOrientation=""
            hoveredAnimation
          />
        </BrowserRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
