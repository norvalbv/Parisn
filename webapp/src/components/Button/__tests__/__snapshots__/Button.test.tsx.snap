// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> Button props with all props as required 1`] = `
<button
  className="relative inline-flex items-center  hover:bg-buttons-hover transition-colors duration-200 uppercase  px-6 py-2.5 text-sm w-full underline border py-4"
  data-testid="Data Attribute"
  disabled={true}
  id="Test Id"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  role="button"
  style={
    {
      "backgroundColor": "bg-red-200",
      "color": "text-red-500",
      "width": "25rem",
    }
  }
  type="button"
>
  <svg
    className="absolute"
    height="100%"
    width="100%"
  >
    <line
      stroke="darkRed"
      strokeWidth={2.25}
      x2="100%"
      y2="100%"
    />
  </svg>
  <span
    className="mr-2.5"
  >
    <div>
      Icon
    </div>
  </span>
  <span
    className="inline-block font-semibold"
  >
    Test Text
  </span>
</button>
`;

exports[`<Button /> Button props with mandatory props 1`] = `
<button
  className="relative inline-flex items-center justify-center hover:bg-buttons-hover transition-colors duration-200 uppercase  px-6 py-2.5 text-sm w-full  border py-4"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  role="button"
  style={
    {
      "backgroundColor": undefined,
      "color": undefined,
      "width": "25rem",
    }
  }
  type="button"
>
  <span
    className="inline-block font-semibold"
  >
    Test Text
  </span>
</button>
`;
